samples/event-sourcing-esdb-simple/src/main/java/io/eventdriven/ecommerce/core/entities/EntityStore.java

```diff
@@ -0,0 +1,96 @@
+package io.eventdriven.ecommerce.core.entities;
+
+import com.eventstore.dbclient.AppendToStreamOptions;
+import com.eventstore.dbclient.EventStoreDBClient;
+import com.eventstore.dbclient.ExpectedRevision;
+import io.eventdriven.ecommerce.core.http.ETag;
+import io.eventdriven.ecommerce.core.serialization.EventSerializer;
+
+import java.util.Optional;
+import java.util.UUID;
+import java.util.concurrent.ExecutionException;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+public class EntityStore<Entity> {
+  private final EventStoreDBClient eventStore;
+  private final BiFunction<Entity, Object, Entity> when;
+  private final Function<UUID, String> mapToStreamId;
+  private final Supplier<Entity> getDefault;
+
+  public EntityStore(
+    EventStoreDBClient eventStore,
+    BiFunction<Entity, Object, Entity> when,
+    Function<UUID, String> mapToStreamId,
+    Supplier<Entity> getDefault
+  ) {
+
+    this.eventStore = eventStore;
+    this.when = when;
+    this.mapToStreamId = mapToStreamId;
+    this.getDefault = getDefault;
+  }
+
+  public Entity get(UUID id) throws ExecutionException, InterruptedException {
+    var streamId = mapToStreamId.apply(id);
+    var result = eventStore.readStream(streamId).get();
+
+    var events = result.getEvents().stream()
+      .map(e -> EventSerializer.deserialize(e))
+      .toList();
+
+    var current = getDefault.get();
+    for (var event : events) {
+      current = when.apply(current, event);
+    }
+
+    return current;
+  }
+
+  public ETag add(
+    Supplier<Object> handle,
+    UUID id
+  ) throws ExecutionException, InterruptedException {
```

Comment:
In Java there are both Checked and Unchecked Exceptions. See more [here](https://baeldung-cn.com/java-checked-unchecked-exceptions).

With CheckedExceptions you have to declare them in every method that uses the method that has CheckedExceptions. To avoid that you can add a try/catch statement and throw a UncheckedException. Then you can have a [GlobalExceptionHandler](https://www.baeldung.com/java-global-exception-handler) so you return the correct message to the client.

Reply:
Good hint, I refactored Exception handling to send only checked exceptions where I intend to handle them in code. I also added ControllerAdvice to map them to proper HTTP statuses: https://github.com/oskardudycz/EventSourcing.JVM/pull/1/commits/ec0198125a019daa9bdfdc574e7cc51ef7761d6e.